src/uri/punycode.cr

```diff
@@ -0,0 +1,181 @@
+# `Punycode` provides an interface for IDNA encoding (RFC 5980),
+# which is defined in RFC 3493
+#
+# Implementation based on Mathias Bynens `punnycode.js` project
+# https://github.com/bestiejs/punycode.js/
+#
+# RFC 3492:
+# Method to use non-ascii characters as host name of URI
+# https://www.ietf.org/rfc/rfc3492.txt
+#
+# RFC 5980:
+# Internationalized Domain Names in Application
+# https://www.ietf.org/rfc/rfc5980.txt
+class URI
+  class Punycode
+    private BASE         =  36
+    private TMIN         =   1
+    private TMAX         =  26
+    private SKEW         =  38
+    private DAMP         = 700
+    private INITIAL_BIAS =  72
+    private INITIAL_N    = 128
+
+    private DELIMITER = '-'
+
+    private BASE36 = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+    private def self.adapt(delta, numpoints, firsttime)
+      delta /= firsttime ? DAMP : 2
+      delta += delta / numpoints
+      k = 0
+      while delta > ((BASE - TMIN) * TMAX) / 2
+        delta /= BASE - TMIN
+        k += BASE
+      end
+      k + (((BASE - TMIN + 1) * delta) / (delta + SKEW))
+    end
+
+    def self.encode(string)
+      String.build { |io| encode string, io }
+    end
+
+    def self.encode(string, io)
+      h = 0
+      all = true
+      others = [] of Char
+
+      string.each_char do |c|
+        if c < '\u0080'
+          h += 1
+          io << c
+          all = false
+        else
+          others.push c
+        end
+      end
+
+      return if others.empty?
+      others.sort!
+      io << DELIMITER unless all
+
+      delta = 0_u32
+      n = INITIAL_N
+      bias = INITIAL_BIAS
+      firsttime = true
+      prev = nil
+
+      h += 1
```

Comment:
You could probably set the value to `h = string.size - others.size + 1` instead of incrementing it in the loop (just a minor improvement).